pipelines/tekton/pipeplines_provider.go

```diff
@@ -0,0 +1,199 @@
+package tekton
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/tektoncd/cli/pkg/pipelinerun"
+	"github.com/tektoncd/cli/pkg/taskrun"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	fn "knative.dev/kn-plugin-func"
+	"knative.dev/kn-plugin-func/k8s"
+	"knative.dev/kn-plugin-func/knative"
+	"knative.dev/kn-plugin-func/pipelines"
+	"knative.dev/pkg/apis"
+)
+
+type Opt func(*PipelinesProvider) error
+
+type PipelinesProvider struct {
+	// namespace with which to override that set on the default configuration (such as the ~/.kube/config).
+	// If left blank, pipeline creation/run will commence to the configured namespace.
+	namespace                            string
+	Verbose                              bool
+	progressListener                     fn.ProgressListener
+	containerRegistryCredentialsCallback pipelines.ContainerRegistryCredentialsCallback
+}
+
+func WithNamespace(namespace string) Opt {
+	return func(pp *PipelinesProvider) error {
+		namespace, err := k8s.GetNamespace(namespace)
+		if err != nil {
+			return err
+		}
+		pp.namespace = namespace
+		return nil
+	}
+}
+
+func WithProgressListener(pl fn.ProgressListener) Opt {
+	return func(pp *PipelinesProvider) error {
+		pp.progressListener = pl
+		return nil
+	}
+}
+
+func WithPromptForContainerRegistryCredentials(cbk pipelines.ContainerRegistryCredentialsCallback) Opt {
+	return func(pp *PipelinesProvider) error {
+		pp.containerRegistryCredentialsCallback = cbk
+		return nil
+	}
+}
+
+func NewPipelinesProvider(opts ...Opt) (*PipelinesProvider, error) {
+	pp := &PipelinesProvider{}
+
+	for _, opt := range opts {
+		err := opt(pp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pp, nil
+}
+
+func (pp *PipelinesProvider) Run(ctx context.Context, f fn.Function) error {
```

Comment:
Some descriptive comments here might be helpful - in general, there is a lot going on in this file. I think comments throughout could be beneficial.